libpgcommon/lwgeom_cache.c

```diff
@@ -236,4 +239,64 @@ GetGeomCache(FunctionCallInfo fcinfo,
 	return NULL;
 }
 
+/******************************************************************************/
 
+inline static ToastCache*
+ToastCacheGet(FunctionCallInfo fcinfo)
+{
+	const uint32_t entry_number = TOAST_CACHE_ENTRY;
+	GenericCacheCollection* generic_cache = GetGenericCacheCollection(fcinfo);
+	ToastCache* cache = (ToastCache*)(generic_cache->entry[entry_number]);
+	if (!cache)
+	{
+		cache = MemoryContextAllocZero(FIContext(fcinfo), sizeof(ToastCache));
+		cache->type = entry_number;
+		generic_cache->entry[entry_number] = (GenericCache*)cache;
+	}
+	return cache;
+}
+
+GSERIALIZED*
+ToastCacheGetGeometry(FunctionCallInfo fcinfo, uint32_t argnum)
+{
+	Assert(argnum < ToastCacheSize);
+	ToastCache* cache = ToastCacheGet(fcinfo);
+	ToastCacheArgument* arg = &(cache->arg[argnum]);
+
+	Datum datum = PG_GETARG_DATUM(argnum);
+	struct varlena *attr = (struct varlena *) DatumGetPointer(datum);
+
+	/* Argument is not TOASTED, access as normal. */
+	if (!VARATT_IS_EXTERNAL_ONDISK(attr))
```

Comment:
PG's `pg_detoast_datum` will do work with anything coming from disk (`VARATT_IS_EXTERNAL_ONDISK`), but also indirect access (`VARATT_IS_EXTERNAL_INDIRECT`), expanded (`VARATT_IS_EXTERNAL_EXPANDED`) and so on; so I'm curious about why just limit the cache to data coming from disk and not to all extended objects (matching detoast).

Reply:
This is following the Word of Tom, and probably he could elabourate, but the key is that when toasted on disk, an object has a relid and a varid that are unique. Before then it'll have a pointer into a memory buffer that *might* change underneath it. So only things that have been stored on disk can be identified uniquely without having to first read in the contents of the object. Fortunately, that's most of the things we care about.

Reply:
Can you leave a comment about it please? I'm bound to think about this next time I look at the code a few months from now.